rfcs/20190729-keras-preprocessing-redesign.md

```diff
@@ -0,0 +1,351 @@
+
+```python
+Resizing(height, width)  # Resize while distorting aspect ratio
+CenterCrop(height, width)  # Resize without distorting aspect ratio
+Rescaling(value)  # Divide by `value`
+RandomFlip(horizontal=False, vertical=False, seed=None)
+RandomTranslation(amplitude=0., fill_mode='constant', fill_value=0., seed=None)
+RandomRotation(amplitude=0., fill_mode='constant', fill_value=0., seed=None)
+RandomZoom(amplitude=0., fill_mode='constant', fill_value=0., seed=None)
+RandomBrightness(amplitude=0., seed=None)
+RandomContrast(amplitude=0., seed=None)
+RandomSaturation(amplitude=0., seed=None)
+RandomWidth(amplitude=0., seed=None)  # Expand / shrink width while distorting aspect ratio
+RandomHeight(amplitude=0., seed=None)  # Expand / shrink height while distorting aspect ratio
+```
+
+The `amplitude` argument may be:
+- a positive float: it is understood as "fraction of total" (total is the current width, or height, or 180 degrees in the case `RandomRotation`). E.g. `0.2` results in variations in the [-20%, +20%] range. If larger than 1, it is rounded to one for the lower boundary (but not the higher boundary).
+- a tuple of 2 positive floats: understood as a fractional range, e.g. `(0.2, 0.4)` is interpreted as the [-20%, +40%] range. The first float may not be larger than 1.
+
+To do a random center crop that zooms in and discards part of the image, you would do:
+
+```python
+preprocessor = ImagePipeline([
+  RandomZoom([0., 0.2]),
+  CenterCrop(height, width),
+])
+```
+
+
+#### Notes
+
+- We are dropping support for ZCA whitening as it is no longer popular in the computer vision community.
+- We don't have immediate support for random translations along only one axis.
+- We only plan on implementing support for `data_format='channels_last'`. As such this argument does not appear in the API.
+
+
+#### Example implementation
+
+```python
+class RandomFlip(PreprocessingLayer):
+
+  def __init__(self, horizontal=False, vertical=False, seed=None):
+    self.horizontal = horizontal
+    self.vertical = vertical
+    self.seed = seed
+    self._current_seed = seed or random_value()
+
+  def call(self, inputs, training=None):
+    seed = self._current_seed
+    self._current_seed += 1
+    if training:
+      if self.horizontal:
+        inputs = tf.image.random_flip_left_right(inputs, seed=seed)
+      if self.vertical:
+        inputs = tf.image.random_flip_up_down(inputs, seed=seed)
+    return inputs
+```
+
+
+
+#### Question: how to support image segmentation in a simple way?
+
+**Requirements:**
+- Image loading and image augmentation should be synced across inputs and targets
+- It should be possible to use different standardization preprocessing (outside of augmentation) across inputs and targets
+
+**Proposal:**
+
+```python
+# Shared spatial transformations for inputs and targets
+augmenter = ImagePipeline([
+    RandomRotation(0.5),
+    RandomFlip(vertical=True)
+])
+
+input_pipeline = ImagePipeline([
+    augmenter,
```

Comment:
We'll need to take a lot of care with the `random` augmentation layers to make sure the syncing works correctly.  They would need to be deterministic & 're-settable' in some fashion that's built-in to the API & that the ImagePipeline/`from_directory` apis would take advantage of.

Reply:
And to make matters trickier now that I think about it, this would also need to work well w/ parallel processing in the dataset (& possibly w/ distribution strategies active). It may be worth taking a page out of Jax's book as inspiration? https://github.com/google/jax/blob/master/design_notes/prng.md

Reply:
Right, we should add a public API to control the seeding behavior, besides the `seed` argument in the constructor.

Would you have a specific API in mind? (in terms of methods and their signature)

Reply:
Hmm, I think the safest thing would be to make random augmentation layers only use stateless random ops and support taking a `seed` argument directly in the call methods. The constructor-time seed argument would then be renamed `initial_seed`. If no `seed` argument is provided at call-time, the seed generated from `initial_seed` would be used.

So, the implementation & API for RandomFlip would look something like:
```python
class RandomFlip(PreprocessingLayer):
   def __init__(self, horizontal=False, vertical=False, initial_seed=None):
    self.horizontal = horizontal
    self.vertical = vertical
    self.initial_seed = initial_seed
    self._current_seed = initial_seed or random_value()
   def call(self, inputs, training=None, seed=None):
    if seed is None:
      seed = self._current_seed
      self._current_seed += 1
    if training:
      if self.horizontal:
        inputs = tf.image.random_flip_left_right(inputs, seed=seed)
      if self.vertical:
        inputs = tf.image.random_flip_up_down(inputs, seed=seed)
    return inputs
```

We'll run into similar challenges as with the `training` argument of layers & models where we have to feed it through nested models & layers to avoid bugs. We can use a similar mechanism in `__call__` to solve the problem.

The one extra thing we would need to do that we don't for the `training` arg is: we'll have to feed each layer in the model a unique seed rather than feeding them all with the same value.

This sort of deterministic randomness could be generally useful for random models & layers in Keras beyond just for random augmentations & preprocessing layers (e.g. for dropout layers)

---------------

The `from_directory` & input methods would continue to take an optional `seed` argument.  They could either provide the dataset tuple index as a seed in the 'call' method of ImagePIpeline, or use some sort of 'tf.function & distribution strategy-friendly' version of `tf.random.experimental.Generator` to generate the seeds for the layers. We can check with Peng who's been working on it to see what the status there is.

The random layers should also use tf.random.experimental.Generator or something similar to maintain their internal seeds rather than using raw python in the form of `self.seed = self.seed + 1`.
That way they will work correctly w/ tf.function & distribution strategies)

Reply:
+1 to these suggestions, especially having `seed` as a `call` argument. Note that we don't need to change the name of the constructor argument; `seed` is fine.

Reply:
So the workflow for an image-segmentation pipeline would be:

Create two identical `ImagePipeline`s, with the same seed, and run them on both inputs? I guess the alternatives would be: 

  - Make these work on `nests` of images
  - Or have a `.reapply` method? 

---------------
So if someone wanted to use this for **object detection** then you could build a `BBox` equivalent of each layer, and write a converter that makes a new `Sequential` pipeline with all the `BBox` layers substituted in, using the same seed. Would that be a reasonable approach to using this for  object_detection?